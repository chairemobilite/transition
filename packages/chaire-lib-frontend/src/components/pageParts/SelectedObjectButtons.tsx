/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { withTranslation, WithTranslation } from 'react-i18next';
import { faArrowLeft } from '@fortawesome/free-solid-svg-icons/faArrowLeft';
import { faUndoAlt } from '@fortawesome/free-solid-svg-icons/faUndoAlt';
import { faRedoAlt } from '@fortawesome/free-solid-svg-icons/faRedoAlt';
import { faTrashAlt } from '@fortawesome/free-solid-svg-icons/faTrashAlt';
import { faCheckCircle } from '@fortawesome/free-solid-svg-icons/faCheckCircle';

import Button from '../input/Button';
import serviceLocator from 'chaire-lib-common/lib/utils/ServiceLocator';
import { ObjectWithHistory } from 'chaire-lib-common/lib/utils/objects/ObjectWithHistory';
import { Saveable, isSaveable } from 'chaire-lib-common/lib/utils/objects/Saveable';

export type SelectedObjectButtonsProps<T extends ObjectWithHistory<any>> = WithTranslation & {
    object: T;
    /**
     * Function called when the save button is clicked. It replaces the default
     * save action, which validates the object and calls the save method on it
     * if there was changes. It also deselects the current object
     */
    saveAction?: React.MouseEventHandler;
    /**
     * Function called after the object has been saved with the default save
     * function. If saveAction is set, then this will have no effect
     */
    afterSave?: (param: any) => void;
    /**
     * Function called after the undo button was clicked and the last action was
     * undone
     */
    onUndo?: (object: T) => void;
    /**
     * Function called after the redo button was clicked and the last action was
     * redone
     */
    onRedo?: (object: T) => void;
    /**
     * Whether to hide or show the delete button
     */
    hideDelete?: boolean;
    /**
     * Whether to hide or show the save button (useful when errors are detected in forms)
     */
    hideSave?: boolean;
    /**
     * Function called when the delete button is clicked and the object is new,
     * otherwise, it is the openDeleteConfirmModal that is called
     */
    deleteAction?: React.MouseEventHandler;
    // @deprecated Use deleteAction instead, for coherence with saveAction
    onDelete?: React.MouseEventHandler;
    // TODO Should this be part of the onDelete on the caller? The code uses it
    // only when the object already exists. Should this become a stateful
    // component and take care of the delete confirm modal?
    openDeleteConfirmModal?: React.MouseEventHandler;

    backAction?: React.MouseEventHandler;
    openBackConfirmModal?: React.MouseEventHandler;
};

const SelectedObjectButtons: React.FunctionComponent<SelectedObjectButtonsProps<any>> = <
    T extends ObjectWithHistory<any>
>(
        props: SelectedObjectButtonsProps<T>
    ) => {
    const object = props.object;
    const objectSingularName = object.getSingularName();
    const objectPluralName = object.getPluralName();
    const objectCapitalizedSingularName = object.getCapitalizedSingularName();
    const isFrozen = object.getAttributes().is_frozen;
    // That may not be the role of this widget to deal with this manager
    const selectObjectManager = serviceLocator.selectedObjectsManager;
    const deleteAction = props.deleteAction ? props.deleteAction : props.onDelete;
    const saveAction = props.saveAction
        ? props.saveAction
        : function () {
            if (!isSaveable(object)) {
                console.error(
                    'The object to save is not saveable. The default save will not work on them. You may specify a saveAction props to customize how to save the object'
                );
                return;
            }
            if (isFrozen === true && object.wasFrozen()) {
                serviceLocator.selectedObjectsManager.deselect(objectSingularName);
                return true;
            }
            if (object.validate()) {
                if (object.hasChanged()) {
                    serviceLocator.eventManager.emit('progress', {
                        name: `Saving${objectCapitalizedSingularName}`,
                        progress: 0.0
                    });
                    object
                        .save(serviceLocator.socketEventManager)
                        .then((response) => {
                            if (props.afterSave) {
                                props.afterSave(response);
                            }
                            serviceLocator.selectedObjectsManager.deselect(objectSingularName);
                            serviceLocator.collectionManager.refresh(objectPluralName);
                            serviceLocator.eventManager.emit('progress', {
                                name: `Saving${objectCapitalizedSingularName}`,
                                progress: 1.0
                            });
                        })
                        .catch((error) => {
                            console.error(error); // todo: better error handling
                        });
                } else {
                    if (selectObjectManager) serviceLocator.selectedObjectsManager.deselect(objectSingularName);
                }
            } else {
                if (selectObjectManager) serviceLocator.selectedObjectsManager.update(objectSingularName, object);
            }
        };

    const undoClick = () => {
        object.undo();
        if (props.onUndo) {
            props.onUndo(object);
        }
        if (selectObjectManager) serviceLocator.selectedObjectsManager.update(objectSingularName, object);
    };

    const redoClick = () => {
        object.redo();
        if (props.onRedo) {
            props.onRedo(object);
        }
        if (selectObjectManager) serviceLocator.selectedObjectsManager.update(objectSingularName, object);
    };

    return (
        <div className="tr__form-buttons-container tr__form-selected-object-buttons-container">
            <Button
                title={props.t('main:Back')}
                name="back"
                key="back"
                color="blue"
                icon={faArrowLeft}
                iconClass="_icon-alone"
                label=""
                onClick={object.hasChanged() ? props.openBackConfirmModal : props.backAction}
            />
            {isFrozen !== true && (
                <Button
                    title={props.t('main:Undo')}
                    name="undo"
                    key="undo"
                    color="grey"
                    disabled={!object.canUndo()}
                    icon={faUndoAlt}
                    iconClass="_icon-alone"
                    label=""
                    onClick={undoClick}
                />
            )}
            {isFrozen !== true && (
                <Button
                    title={props.t('main:Redo')}
                    name="redo"
                    key="redo"
                    color="grey"
                    disabled={!object.canRedo()}
                    icon={faRedoAlt}
                    iconClass="_icon-alone"
                    label=""
                    onClick={redoClick}
                />
            )}
            {props.hideSave !== true && (
                <Button
                    title={props.t('main:Save')}
                    name="save"
                    key="save"
                    icon={faCheckCircle}
                    iconClass="_icon-alone"
                    label=""
                    onClick={saveAction}
                />
            )}
            {isFrozen !== true && props.hideDelete !== true && (
                <Button
                    title={props.t('main:Delete')}
                    name="delete"
                    key="delete"
                    color="red"
                    icon={faTrashAlt}
                    iconClass="_icon-alone"
                    label=""
                    onClick={object.isNew() ? deleteAction : props.openDeleteConfirmModal}
                />
            )}
        </div>
    );
};

export default withTranslation(['main', 'notifications'])(SelectedObjectButtons);
