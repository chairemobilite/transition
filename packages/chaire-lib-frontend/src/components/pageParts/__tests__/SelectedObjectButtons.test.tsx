/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import * as React from 'react';

import SelectedObjectButtons from '../SelectedObjectButtons';
import { ObjectWithHistory } from 'chaire-lib-common/lib/utils/objects/ObjectWithHistory';
import { GenericAttributes } from 'chaire-lib-common/lib/utils/objects/GenericObject';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

// TODO Cannot do a stub class, it gives compilation error. Wait for a mocking/stubbing library
let newObject: ObjectWithHistory<GenericAttributes>;
let existingObject: ObjectWithHistory<GenericAttributes>;

beforeEach(() => {
    existingObject = new ObjectWithHistory({
        id: 'abcdef'
    }, false);
    existingObject.startEditing();
    newObject = new ObjectWithHistory({}, true);
    newObject.startEditing();
});

test('Test with delete button', () => {
    const { container } = render(<SelectedObjectButtons
        object = {existingObject}
    />);
    expect(container).toMatchSnapshot();
});

test('Test without delete button', () => {
    const { container } = render(<SelectedObjectButtons
        object = {existingObject}
        hideDelete = {true}
    />);
    expect(container).toMatchSnapshot();
});

test('Test with save button', () => {
    const { container } = render(<SelectedObjectButtons
        object = {existingObject}
    />);
    expect(container).toMatchSnapshot();
});

test('Test without save button', () => {
    const { container } = render(<SelectedObjectButtons
        object = {existingObject}
        hideSave = {true}
    />);
    expect(container).toMatchSnapshot();
});

test('Test back click on unmodified object', () => {
    const mockBackAction = jest.fn();
    const mockOpenModal = jest.fn();
    render(<SelectedObjectButtons
        object = {existingObject}
        backAction = {mockBackAction}
        openBackConfirmModal = {mockOpenModal}
    />);

    const backButton = screen.getByTitle('main:Back');
    fireEvent.click(backButton);

    expect(mockBackAction).toHaveBeenCalledTimes(1);
    expect(mockOpenModal).toHaveBeenCalledTimes(0);
});

test('Test back click on modified object', () => {
    const mockBackAction = jest.fn();
    const mockOpenModal = jest.fn();

    existingObject.setData('modification', 1234);
    /*
    With enzyme, we can change the object after render, but for a weird reason,
    with testing-library, we cannot. maybe the rendering just clones the atributes.
    So we need to do the change before the render.
    */
    render(<SelectedObjectButtons
        object={existingObject}
        backAction={mockBackAction}
        openBackConfirmModal={mockOpenModal}
    />);

    const backButton = screen.getByTitle('main:Back');
    fireEvent.click(backButton);
    console.log('hasCHanged');

    expect(mockOpenModal).toHaveBeenCalledTimes(1);
    expect(mockBackAction).toHaveBeenCalledTimes(0);
});

test('Test undo click', () => {
    const mockUndo = jest.fn();
    const { rerender } = render(<SelectedObjectButtons
        object = {existingObject}
        onUndo = {mockUndo}
    />);

    // Should be disabled now
    const undoButton = screen.getByTitle('main:Undo');
    expect(undoButton).toBeDisabled();

    // Do a change on the object and set the props again
    const dataToUndo = 'toUndo';
    const value = 1234;
    existingObject.startEditing();
    existingObject.setData(dataToUndo, value);
    rerender(<SelectedObjectButtons
        object={existingObject}
        onUndo={mockUndo}
    />);
    expect(undoButton).not.toBeDisabled();
    fireEvent.click(undoButton);
    expect(mockUndo).toHaveBeenCalledTimes(1);
    expect(existingObject.getData(dataToUndo)).toBeUndefined();
    existingObject.stopEditing();
});

test('Test redo click', () => {
    const mockRedo = jest.fn();
    const { rerender } = render(<SelectedObjectButtons
        object = {existingObject}
        onRedo = {mockRedo}
    />);

    // Should be disabled now
    const redoButton = screen.getByTitle('main:Redo');
    expect(redoButton).toBeDisabled();

    // Do a change on the object and set the props again
    const dataToUndo = 'toUndo';
    const value = 1234;
    existingObject.startEditing();
    existingObject.setData(dataToUndo, value);
    existingObject.undo();
    rerender(<SelectedObjectButtons
        object={existingObject}
        onRedo={mockRedo}
    />);
    expect(redoButton).not.toBeDisabled();
    fireEvent.click(redoButton);
    expect(mockRedo).toHaveBeenCalledTimes(1);
    expect(existingObject.getData(dataToUndo)).toEqual(value);
    existingObject.stopEditing();
});

test('Test default save click', () => {

    // TODO Wait for stubbing library for easier test
});

test('Test custom save action', () => {

    // TODO Wait for stubbing library for easier test
});

test('Test delete for new objects', () => {
    const mockDeleteAction = jest.fn();
    const mockOpenModal = jest.fn();
    render(<SelectedObjectButtons
        object={newObject}
        onDelete={mockDeleteAction}
        openDeleteConfirmModal={mockOpenModal}
    />);

    const deleteButton = screen.getByTitle('main:Delete');
    fireEvent.click(deleteButton);

    expect(mockDeleteAction).toHaveBeenCalledTimes(1);
    expect(mockOpenModal).toHaveBeenCalledTimes(0);
});

test('Test delete for existing objects', () => {
    const mockDeleteAction = jest.fn();
    const mockOpenModal = jest.fn();

    render(<SelectedObjectButtons
        object={existingObject}
        onDelete={mockDeleteAction}
        openDeleteConfirmModal={mockOpenModal}
    />);

    const deleteButton = screen.getByTitle('main:Delete');
    fireEvent.click(deleteButton);

    expect(mockOpenModal).toHaveBeenCalledTimes(1);
    expect(mockDeleteAction).toHaveBeenCalledTimes(0);
});
