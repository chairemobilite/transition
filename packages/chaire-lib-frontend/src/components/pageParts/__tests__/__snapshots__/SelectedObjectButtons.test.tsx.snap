// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test with delete button 1`] = `
<div>
  <div
    class="tr__form-buttons-container tr__form-selected-object-buttons-container"
  >
    <div
      class="center"
    >
      <button
        class="button blue large"
        name="back"
        title="main:Back"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-arrow-left _icon-alone"
          data-icon="arrow-left"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.2 288 416 288c17.7 0 32-14.3 32-32s-14.3-32-32-32l-306.7 0L214.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"
            fill="currentColor"
          />
        </svg>
         
      </button>
    </div>
    <div
      class="center"
    >
      <button
        class="button grey large disabled"
        disabled=""
        name="undo"
        title="main:Undo"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-rotate-left _icon-alone"
          data-icon="rotate-left"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M48.5 224L40 224c-13.3 0-24-10.7-24-24L16 72c0-9.7 5.8-18.5 14.8-22.2s19.3-1.7 26.2 5.2L98.6 96.6c87.6-86.5 228.7-86.2 315.8 1c87.5 87.5 87.5 229.3 0 316.8s-229.3 87.5-316.8 0c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0c62.5 62.5 163.8 62.5 226.3 0s62.5-163.8 0-226.3c-62.2-62.2-162.7-62.5-225.3-1L185 183c6.9 6.9 8.9 17.2 5.2 26.2s-12.5 14.8-22.2 14.8L48.5 224z"
            fill="currentColor"
          />
        </svg>
         
      </button>
    </div>
    <div
      class="center"
    >
      <button
        class="button grey large disabled"
        disabled=""
        name="redo"
        title="main:Redo"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-rotate-right _icon-alone"
          data-icon="rotate-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M463.5 224l8.5 0c13.3 0 24-10.7 24-24l0-128c0-9.7-5.8-18.5-14.8-22.2s-19.3-1.7-26.2 5.2L413.4 96.6c-87.6-86.5-228.7-86.2-315.8 1c-87.5 87.5-87.5 229.3 0 316.8s229.3 87.5 316.8 0c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0c-62.5 62.5-163.8 62.5-226.3 0s-62.5-163.8 0-226.3c62.2-62.2 162.7-62.5 225.3-1L327 183c-6.9 6.9-8.9 17.2-5.2 26.2s12.5 14.8 22.2 14.8l119.5 0z"
            fill="currentColor"
          />
        </svg>
         
      </button>
    </div>
    <div
      class="center"
    >
      <button
        class="button green large"
        name="save"
        title="main:Save"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-circle-check _icon-alone"
          data-icon="circle-check"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM369 209L241 337c-9.4 9.4-24.6 9.4-33.9 0l-64-64c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.4 33.9 0l47 47L335 175c9.4-9.4 24.6-9.4 33.9 0s9.4 24.6 0 33.9z"
            fill="currentColor"
          />
        </svg>
         
      </button>
    </div>
    <div
      class="center"
    >
      <button
        class="button red large"
        name="delete"
        title="main:Delete"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-trash-can _icon-alone"
          data-icon="trash-can"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M135.2 17.7C140.6 6.8 151.7 0 163.8 0L284.2 0c12.1 0 23.2 6.8 28.6 17.7L320 32l96 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 96C14.3 96 0 81.7 0 64S14.3 32 32 32l96 0 7.2-14.3zM32 128l384 0 0 320c0 35.3-28.7 64-64 64L96 512c-35.3 0-64-28.7-64-64l0-320zm96 64c-8.8 0-16 7.2-16 16l0 224c0 8.8 7.2 16 16 16s16-7.2 16-16l0-224c0-8.8-7.2-16-16-16zm96 0c-8.8 0-16 7.2-16 16l0 224c0 8.8 7.2 16 16 16s16-7.2 16-16l0-224c0-8.8-7.2-16-16-16zm96 0c-8.8 0-16 7.2-16 16l0 224c0 8.8 7.2 16 16 16s16-7.2 16-16l0-224c0-8.8-7.2-16-16-16z"
            fill="currentColor"
          />
        </svg>
         
      </button>
    </div>
  </div>
</div>
`;

exports[`Test with save button 1`] = `
<div>
  <div
    class="tr__form-buttons-container tr__form-selected-object-buttons-container"
  >
    <div
      class="center"
    >
      <button
        class="button blue large"
        name="back"
        title="main:Back"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-arrow-left _icon-alone"
          data-icon="arrow-left"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.2 288 416 288c17.7 0 32-14.3 32-32s-14.3-32-32-32l-306.7 0L214.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"
            fill="currentColor"
          />
        </svg>
         
      </button>
    </div>
    <div
      class="center"
    >
      <button
        class="button grey large disabled"
        disabled=""
        name="undo"
        title="main:Undo"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-rotate-left _icon-alone"
          data-icon="rotate-left"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M48.5 224L40 224c-13.3 0-24-10.7-24-24L16 72c0-9.7 5.8-18.5 14.8-22.2s19.3-1.7 26.2 5.2L98.6 96.6c87.6-86.5 228.7-86.2 315.8 1c87.5 87.5 87.5 229.3 0 316.8s-229.3 87.5-316.8 0c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0c62.5 62.5 163.8 62.5 226.3 0s62.5-163.8 0-226.3c-62.2-62.2-162.7-62.5-225.3-1L185 183c6.9 6.9 8.9 17.2 5.2 26.2s-12.5 14.8-22.2 14.8L48.5 224z"
            fill="currentColor"
          />
        </svg>
         
      </button>
    </div>
    <div
      class="center"
    >
      <button
        class="button grey large disabled"
        disabled=""
        name="redo"
        title="main:Redo"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-rotate-right _icon-alone"
          data-icon="rotate-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M463.5 224l8.5 0c13.3 0 24-10.7 24-24l0-128c0-9.7-5.8-18.5-14.8-22.2s-19.3-1.7-26.2 5.2L413.4 96.6c-87.6-86.5-228.7-86.2-315.8 1c-87.5 87.5-87.5 229.3 0 316.8s229.3 87.5 316.8 0c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0c-62.5 62.5-163.8 62.5-226.3 0s-62.5-163.8 0-226.3c62.2-62.2 162.7-62.5 225.3-1L327 183c-6.9 6.9-8.9 17.2-5.2 26.2s12.5 14.8 22.2 14.8l119.5 0z"
            fill="currentColor"
          />
        </svg>
         
      </button>
    </div>
    <div
      class="center"
    >
      <button
        class="button green large"
        name="save"
        title="main:Save"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-circle-check _icon-alone"
          data-icon="circle-check"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM369 209L241 337c-9.4 9.4-24.6 9.4-33.9 0l-64-64c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.4 33.9 0l47 47L335 175c9.4-9.4 24.6-9.4 33.9 0s9.4 24.6 0 33.9z"
            fill="currentColor"
          />
        </svg>
         
      </button>
    </div>
    <div
      class="center"
    >
      <button
        class="button red large"
        name="delete"
        title="main:Delete"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-trash-can _icon-alone"
          data-icon="trash-can"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M135.2 17.7C140.6 6.8 151.7 0 163.8 0L284.2 0c12.1 0 23.2 6.8 28.6 17.7L320 32l96 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 96C14.3 96 0 81.7 0 64S14.3 32 32 32l96 0 7.2-14.3zM32 128l384 0 0 320c0 35.3-28.7 64-64 64L96 512c-35.3 0-64-28.7-64-64l0-320zm96 64c-8.8 0-16 7.2-16 16l0 224c0 8.8 7.2 16 16 16s16-7.2 16-16l0-224c0-8.8-7.2-16-16-16zm96 0c-8.8 0-16 7.2-16 16l0 224c0 8.8 7.2 16 16 16s16-7.2 16-16l0-224c0-8.8-7.2-16-16-16zm96 0c-8.8 0-16 7.2-16 16l0 224c0 8.8 7.2 16 16 16s16-7.2 16-16l0-224c0-8.8-7.2-16-16-16z"
            fill="currentColor"
          />
        </svg>
         
      </button>
    </div>
  </div>
</div>
`;

exports[`Test without delete button 1`] = `
<div>
  <div
    class="tr__form-buttons-container tr__form-selected-object-buttons-container"
  >
    <div
      class="center"
    >
      <button
        class="button blue large"
        name="back"
        title="main:Back"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-arrow-left _icon-alone"
          data-icon="arrow-left"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.2 288 416 288c17.7 0 32-14.3 32-32s-14.3-32-32-32l-306.7 0L214.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"
            fill="currentColor"
          />
        </svg>
         
      </button>
    </div>
    <div
      class="center"
    >
      <button
        class="button grey large disabled"
        disabled=""
        name="undo"
        title="main:Undo"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-rotate-left _icon-alone"
          data-icon="rotate-left"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M48.5 224L40 224c-13.3 0-24-10.7-24-24L16 72c0-9.7 5.8-18.5 14.8-22.2s19.3-1.7 26.2 5.2L98.6 96.6c87.6-86.5 228.7-86.2 315.8 1c87.5 87.5 87.5 229.3 0 316.8s-229.3 87.5-316.8 0c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0c62.5 62.5 163.8 62.5 226.3 0s62.5-163.8 0-226.3c-62.2-62.2-162.7-62.5-225.3-1L185 183c6.9 6.9 8.9 17.2 5.2 26.2s-12.5 14.8-22.2 14.8L48.5 224z"
            fill="currentColor"
          />
        </svg>
         
      </button>
    </div>
    <div
      class="center"
    >
      <button
        class="button grey large disabled"
        disabled=""
        name="redo"
        title="main:Redo"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-rotate-right _icon-alone"
          data-icon="rotate-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M463.5 224l8.5 0c13.3 0 24-10.7 24-24l0-128c0-9.7-5.8-18.5-14.8-22.2s-19.3-1.7-26.2 5.2L413.4 96.6c-87.6-86.5-228.7-86.2-315.8 1c-87.5 87.5-87.5 229.3 0 316.8s229.3 87.5 316.8 0c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0c-62.5 62.5-163.8 62.5-226.3 0s-62.5-163.8 0-226.3c62.2-62.2 162.7-62.5 225.3-1L327 183c-6.9 6.9-8.9 17.2-5.2 26.2s12.5 14.8 22.2 14.8l119.5 0z"
            fill="currentColor"
          />
        </svg>
         
      </button>
    </div>
    <div
      class="center"
    >
      <button
        class="button green large"
        name="save"
        title="main:Save"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-circle-check _icon-alone"
          data-icon="circle-check"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM369 209L241 337c-9.4 9.4-24.6 9.4-33.9 0l-64-64c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.4 33.9 0l47 47L335 175c9.4-9.4 24.6-9.4 33.9 0s9.4 24.6 0 33.9z"
            fill="currentColor"
          />
        </svg>
         
      </button>
    </div>
  </div>
</div>
`;

exports[`Test without save button 1`] = `
<div>
  <div
    class="tr__form-buttons-container tr__form-selected-object-buttons-container"
  >
    <div
      class="center"
    >
      <button
        class="button blue large"
        name="back"
        title="main:Back"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-arrow-left _icon-alone"
          data-icon="arrow-left"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.2 288 416 288c17.7 0 32-14.3 32-32s-14.3-32-32-32l-306.7 0L214.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"
            fill="currentColor"
          />
        </svg>
         
      </button>
    </div>
    <div
      class="center"
    >
      <button
        class="button grey large disabled"
        disabled=""
        name="undo"
        title="main:Undo"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-rotate-left _icon-alone"
          data-icon="rotate-left"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M48.5 224L40 224c-13.3 0-24-10.7-24-24L16 72c0-9.7 5.8-18.5 14.8-22.2s19.3-1.7 26.2 5.2L98.6 96.6c87.6-86.5 228.7-86.2 315.8 1c87.5 87.5 87.5 229.3 0 316.8s-229.3 87.5-316.8 0c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0c62.5 62.5 163.8 62.5 226.3 0s62.5-163.8 0-226.3c-62.2-62.2-162.7-62.5-225.3-1L185 183c6.9 6.9 8.9 17.2 5.2 26.2s-12.5 14.8-22.2 14.8L48.5 224z"
            fill="currentColor"
          />
        </svg>
         
      </button>
    </div>
    <div
      class="center"
    >
      <button
        class="button grey large disabled"
        disabled=""
        name="redo"
        title="main:Redo"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-rotate-right _icon-alone"
          data-icon="rotate-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M463.5 224l8.5 0c13.3 0 24-10.7 24-24l0-128c0-9.7-5.8-18.5-14.8-22.2s-19.3-1.7-26.2 5.2L413.4 96.6c-87.6-86.5-228.7-86.2-315.8 1c-87.5 87.5-87.5 229.3 0 316.8s229.3 87.5 316.8 0c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0c-62.5 62.5-163.8 62.5-226.3 0s-62.5-163.8 0-226.3c62.2-62.2 162.7-62.5 225.3-1L327 183c-6.9 6.9-8.9 17.2-5.2 26.2s12.5 14.8 22.2 14.8l119.5 0z"
            fill="currentColor"
          />
        </svg>
         
      </button>
    </div>
    <div
      class="center"
    >
      <button
        class="button red large"
        name="delete"
        title="main:Delete"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-trash-can _icon-alone"
          data-icon="trash-can"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M135.2 17.7C140.6 6.8 151.7 0 163.8 0L284.2 0c12.1 0 23.2 6.8 28.6 17.7L320 32l96 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 96C14.3 96 0 81.7 0 64S14.3 32 32 32l96 0 7.2-14.3zM32 128l384 0 0 320c0 35.3-28.7 64-64 64L96 512c-35.3 0-64-28.7-64-64l0-320zm96 64c-8.8 0-16 7.2-16 16l0 224c0 8.8 7.2 16 16 16s16-7.2 16-16l0-224c0-8.8-7.2-16-16-16zm96 0c-8.8 0-16 7.2-16 16l0 224c0 8.8 7.2 16 16 16s16-7.2 16-16l0-224c0-8.8-7.2-16-16-16zm96 0c-8.8 0-16 7.2-16 16l0 224c0 8.8 7.2 16 16 16s16-7.2 16-16l0-224c0-8.8-7.2-16-16-16z"
            fill="currentColor"
          />
        </svg>
         
      </button>
    </div>
  </div>
</div>
`;
